File: typescript/standard/03-job-filtering.ts
/**
 * Standard API Example 03: Job Filtering
 *
 * Demonstrates how to filter which jobs your agent accepts.
 * Useful for specializing in certain types of work.
 *
 * This example demonstrates:
 * - Budget filters (minBudget, maxBudget)
 * - Custom filter functions
 * - Combining filters with pricing
 *
 * Run: npx ts-node standard/03-job-filtering.ts
 */

import { Agent, request, serviceDirectory } from '@agirails/sdk';

async function main() {
  console.log('AGIRAILS Standard API - Job Filtering\n');

  // Create agent
  const agent = new Agent({
    name: 'FilteringAgent',
    network: 'mock',
  });

  // Service 1: With budget filters
  agent.provide(
    {
      name: 'premium-analysis',
      description: 'Premium data analysis service',
      filter: {
        minBudget: 10, // Minimum $10
        maxBudget: 1000, // Maximum $1000
      },
    },
    async (job) => {
      console.log(`  Processing premium analysis (budget: $${job.budget})`);
      return { analysis: 'Premium insights', quality: 'high' };
    }
  );

  // Service 2: With custom filter function
  agent.provide(
    {
      name: 'image-processing',
      description: 'Image processing with size limits',
      filter: {
        // Custom filter: only accept images under 10MB
        custom: (job) => {
          const sizeInMB = job.input.sizeInBytes / (1024 * 1024);
          if (sizeInMB > 10) {
            console.log(`  Rejected: Image too large (${sizeInMB.toFixed(1)}MB)`);
            return false;
          }
          return true;
        },
      },
    },
    async (job) => {
      console.log(`  Processing image (${(job.input.sizeInBytes / 1024).toFixed(0)}KB)`);
      return { processed: true, format: job.input.format };
    }
  );

  // Service 3: Complex filter with multiple conditions
  agent.provide(
    {
      name: 'ai-coding',
      description: 'AI coding assistance',
      filter: {
        minBudget: 5,
        custom: (job) => {
          // Only accept certain programming languages
          const supportedLangs = ['typescript', 'javascript', 'python', 'solidity'];
          const lang = job.input.language?.toLowerCase();
          if (!supportedLangs.includes(lang)) {
            console.log(`  Rejected: Unsupported language (${lang})`);
            return false;
          }
          return true;
        },
      },
    },
    async (job) => {
      console.log(`  Generating ${job.input.language} code...`);
      return { code: `// ${job.input.language} code here`, language: job.input.language };
    }
  );

  await agent.start();
  // Register services with service directory
  serviceDirectory.register('premium-analysis', agent.address);
  serviceDirectory.register('image-processing', agent.address);
  serviceDirectory.register('ai-coding', agent.address);
  console.log('Filtering agent started:', agent.address);
  console.log('Services:', agent.serviceNames.join(', '));
  console.log('');

  // Test premium-analysis filters
  console.log('=== Testing premium-analysis ===\n');

  console.log('Test 1: $5 budget (below minimum $10)');
  try {
    await request('premium-analysis', { input: {}, budget: 5, timeout: 3000 });
    console.log('  Status: ACCEPTED');
  } catch {
    console.log('  Status: REJECTED (below minBudget)');
  }

  console.log('\nTest 2: $50 budget (within range)');
  try {
    await request('premium-analysis', { input: {}, budget: 50 });
    console.log('  Status: ACCEPTED');
  } catch {
    console.log('  Status: REJECTED');
  }

  // Test image-processing filters
  console.log('\n=== Testing image-processing ===\n');

  console.log('Test 3: 5MB image (within limit)');
  try {
    await request('image-processing', {
      input: { sizeInBytes: 5 * 1024 * 1024, format: 'png' },
      budget: 2,
    });
    console.log('  Status: ACCEPTED');
  } catch {
    console.log('  Status: REJECTED');
  }

  console.log('\nTest 4: 15MB image (above limit)');
  try {
    await request('image-processing', {
      input: { sizeInBytes: 15 * 1024 * 1024, format: 'png' },
      budget: 5,
      timeout: 3000,
    });
    console.log('  Status: ACCEPTED');
  } catch {
    console.log('  Status: REJECTED (too large)');
  }

  // Test ai-coding filters
  console.log('\n=== Testing ai-coding ===\n');

  console.log('Test 5: TypeScript code (supported)');
  try {
    await request('ai-coding', {
      input: { language: 'typescript', task: 'write function' },
      budget: 10,
    });
    console.log('  Status: ACCEPTED');
  } catch {
    console.log('  Status: REJECTED');
  }

  console.log('\nTest 6: Rust code (not supported)');
  try {
    await request('ai-coding', {
      input: { language: 'rust', task: 'write function' },
      budget: 10,
      timeout: 3000,
    });
    console.log('  Status: ACCEPTED');
  } catch {
    console.log('  Status: REJECTED (unsupported language)');
  }

  // Summary
  console.log('\n=== Summary ===');
  console.log(`Jobs completed: ${agent.stats.jobsCompleted}`);
  console.log(`Total earned: $${agent.stats.totalEarned.toFixed(2)}`);

  // Unregister services
  serviceDirectory.unregister('premium-analysis', agent.address);
  serviceDirectory.unregister('image-processing', agent.address);
  serviceDirectory.unregister('ai-coding', agent.address);
  await agent.stop();
  console.log('\nFiltering demo complete!');
}

main().catch(console.error);

File: python/standard/03_job_filtering.py
#!/usr/bin/env python3
"""
Standard API Example 03: Job Filtering

Demonstrates how to filter which jobs your agent accepts.
Useful for specializing in certain types of work.

This example demonstrates:
- Budget filters (min_budget, max_budget)
- Custom filter functions
- Combining filters with pricing

Run: python standard/03_job_filtering.py
"""

import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from src.utils.helpers import clear_mock_state


async def main() -> None:
    print("AGIRAILS Standard API - Job Filtering\n")

    clear_mock_state()

    from agirails.level1 import (
        Agent,
        AgentConfig,
        Job,
        JobContext,
        ServiceFilter,
    )

    # Create agent
    agent = Agent(
        AgentConfig(
            name="FilteringAgent",
            network="mock",
        )
    )

    # Service 1: With budget filters
    premium_filter = ServiceFilter(
        min_budget=10.0,  # Minimum $10
        max_budget=1000.0,  # Maximum $1000
    )

    @agent.provide("premium-analysis", filter=premium_filter)
    async def handle_premium(job: Job, ctx: JobContext) -> dict:
        ctx.log.info(f"Processing premium analysis (budget: ${job.budget})")
        return {"analysis": "Premium insights", "quality": "high"}

    # Service 2: With custom filter function
    def image_size_filter(job: Job) -> bool:
        """Only accept images under 10MB."""
        size_bytes = job.input.get("sizeInBytes", 0) if isinstance(job.input, dict) else 0
        size_mb = size_bytes / (1024 * 1024)
        if size_mb > 10:
            print(f"  [Filter] Rejected: Image too large ({size_mb:.1f}MB)")
            return False
        return True

    image_filter = ServiceFilter(custom=image_size_filter)

    @agent.provide("image-processing", filter=image_filter)
    async def handle_image(job: Job, ctx: JobContext) -> dict:
        size_kb = job.input.get("sizeInBytes", 0) / 1024 if isinstance(job.input, dict) else 0
        ctx.log.info(f"Processing image ({size_kb:.0f}KB)")
        return {"processed": True, "format": job.input.get("format", "unknown")}

    # Service 3: Complex filter with multiple conditions
    def coding_language_filter(job: Job) -> bool:
        """Only accept certain programming languages."""
        supported_langs = ["typescript", "javascript", "python", "solidity"]
        lang = job.input.get("language", "").lower() if isinstance(job.input, dict) else ""
        if lang not in supported_langs:
            print(f"  [Filter] Rejected: Unsupported language ({lang})")
            return False
        return True

    coding_filter = ServiceFilter(
        min_budget=5.0,
        custom=coding_language_filter,
    )

    @agent.provide("ai-coding", filter=coding_filter)
    async def handle_coding(job: Job, ctx: JobContext) -> dict:
        lang = job.input.get("language", "unknown") if isinstance(job.input, dict) else "unknown"
        ctx.log.info(f"Generating {lang} code...")
        return {"code": f"# {lang} code here", "language": lang}

    await agent.start()
    print(f"Filtering agent started: {agent.address[:16]}...")
    print(f"Services: {', '.join(agent.service_names)}")
    print()

    # Demo: Show filter behavior with test cases
    print("=" * 50)
    print("Filter Demonstration")
    print("=" * 50)

    # Test premium-analysis
    print("\n=== Testing premium-analysis ===\n")

    test_budgets = [
        (5.0, "Below minimum ($10)"),
        (50.0, "Within range"),
        (1500.0, "Above maximum ($1000)"),
    ]

    for budget, desc in test_budgets:
        print(f"Test: ${budget:.0f} budget - {desc}")
        if budget < premium_filter.min_budget:
            print("  Result: REJECTED (below min_budget)")
        elif budget > premium_filter.max_budget:
            print("  Result: REJECTED (above max_budget)")
        else:
            print("  Result: ACCEPTED")
        print()

    # Test image-processing
    print("=== Testing image-processing ===\n")

    test_sizes = [
        (5 * 1024 * 1024, "5MB - within limit"),
        (15 * 1024 * 1024, "15MB - above limit"),
    ]

    for size, desc in test_sizes:
        print(f"Test: {desc}")
        # Simulate filter check
        if size / (1024 * 1024) > 10:
            print("  Result: REJECTED (file too large)")
        else:
            print("  Result: ACCEPTED")
        print()

    # Test ai-coding
    print("=== Testing ai-coding ===\n")

    test_langs = [
        ("typescript", 10.0, "TypeScript - supported"),
        ("python", 10.0, "Python - supported"),
        ("rust", 10.0, "Rust - not supported"),
        ("python", 2.0, "Python - budget too low"),
    ]

    for lang, budget, desc in test_langs:
        print(f"Test: {desc}")
        supported = lang in ["typescript", "javascript", "python", "solidity"]
        if budget < coding_filter.min_budget:
            print("  Result: REJECTED (below min_budget)")
        elif not supported:
            print("  Result: REJECTED (unsupported language)")
        else:
            print("  Result: ACCEPTED")
        print()

    # Summary
    print("=" * 50)
    print("Summary")
    print("=" * 50)
    print(f"Services registered: {len(agent.service_names)}")
    print(f"  - premium-analysis: min=$10, max=$1000")
    print(f"  - image-processing: max 10MB file size")
    print(f"  - ai-coding: min=$5, languages=[ts, js, py, sol]")

    await agent.stop()
    print("\nFiltering demo complete!")


if __name__ == "__main__":
    asyncio.run(main())
